### PR TITLE
Support soroban upgrades in overlay

### PR DESCRIPTION
**Pull request by marta-lokhova**
Resolves https://github.com/stellar/stellar-core/issues/3746 + bump overlay version

The change potentially triggers upgrade handling every time the network externalizes a new ledger, automatically adjusting capacities to accommodate the new size. Note that when the capacity is increased, all peer connections need to be notified as well, which is done by sending an additional SEND_MORE message.

This upgrade handling applies only in scenarios where the maximum transaction size increases. No capacity adjustment is needed if the max size is decreased since the overlay can already handle those transactions. However, it's worth noting that values from the config file might become invalidated after the upgrade, as they need to maintain a specific invariance with the max tx size. To address this, I updated the code to automatically configure the values if they weren't specified by the operator. If an operator did provide config values, they will be validated during startup. Since these config options are considered more advanced, it's unlikely that many operators actually use them and instead rely on the default settings.

### CODE REVIEW

File: src/herder/HerderImpl.cpp [hunk 1]
```diff
@@ -1934,9 +1950,69 @@ HerderImpl::restoreUpgrades()
...
+        LedgerTxn ltx(mApp.getLedgerTxnRoot(),
+                      /* shouldUpdateLastModified */ true,
+                      TransactionMode::READ_ONLY_WITHOUT_SQL_TXN);
+        auto conf = mApp.getLedgerManager().getSorobanNetworkConfig(ltx);
+
+        if (conf.txMaxSizeBytes() > mMaxTxSize)
+        {
+            diff = conf.txMaxSizeBytes() - mMaxTxSize;
+        }
+        mMaxTxSize =
```
Comment (mbsdf) [inline src/herder/HerderImpl.cpp:1968]: do we need to update FlowControl.cpp L388

File: src/herder/HerderImpl.cpp [hunk 2]
```diff
@@ -1934,9 +1950,69 @@ HerderImpl::restoreUpgrades()
...
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+    {
+        LedgerTxn ltx(mApp.getLedgerTxnRoot(),
+                      /* shouldUpdateLastModified */ true,
+                      TransactionMode::READ_ONLY_WITHOUT_SQL_TXN);
+        auto conf = mApp.getLedgerManager().getSorobanNetworkConfig(ltx);
+        mMaxTxSize = std::max<uint64_t>(sz, conf.txMaxSizeBytes());
+    }
+#else
+    mMaxTxSize = sz;
```
Comment (mbsdf) [inline src/herder/HerderImpl.cpp:1993]: is this by default MAX_CLASSIC_TX_SIZE_BYTES = 100 * 1024

File: src/herder/HerderImpl.h [hunk 3]
```diff
@@ -127,6 +134,13 @@ class HerderImpl : public Herder
 
     void processSCPQueue();
 
+    uint64_t getMaxClassicTxSize() override;
+    uint64_t
+    getMaxTxSize() const override
+    {
+        return mMaxTxSize;
```
Comment (mbsdf) [inline src/herder/HerderImpl.h:141]: ok

File: src/herder/HerderImpl.cpp [hunk 4]
```diff
@@ -1934,9 +1950,69 @@ HerderImpl::restoreUpgrades()
...
+    // Core will calculate default values automatically
+    bool calculateDefaults = cfg.PEER_FLOOD_READING_CAPACITY_BYTES == 0 &&
+                             cfg.FLOW_CONTROL_SEND_MORE_BATCH_SIZE_BYTES == 0;
+
+    if (!calculateDefaults &&
+        !(cfg.PEER_FLOOD_READING_CAPACITY_BYTES -
+              cfg.FLOW_CONTROL_SEND_MORE_BATCH_SIZE_BYTES >=
+          mMaxTxSize))
+    {
+        std::string msg = fmt::format(
```
Comment (mbsdf) [inline src/herder/HerderImpl.cpp:2006]: just to make sure; if we increase tx size in ledger, is it possible that it can create a config error (condition above) when core is restarted?

File: src/main/Config.cpp [hunk 5]
```diff
@@ -246,8 +247,10 @@ Config::Config() : NODE_SEED(SecretKey::random())
     PEER_FLOOD_READING_CAPACITY = 200;
     FLOW_CONTROL_SEND_MORE_BATCH_SIZE = 40;
 
-    PEER_FLOOD_READING_CAPACITY_BYTES = 300000;
-    FLOW_CONTROL_SEND_MORE_BATCH_SIZE_BYTES = 100000;
+    // If set to 0, calculate automatically (this will be done after application
+    // startup as we need to load soroban configs)
+    PEER_FLOOD_READING_CAPACITY_BYTES = 0;
```
Comment (mbsdf) [inline src/main/Config.cpp:252]: regarding comment in HerderImp.cpp L2000 condition;  only if calculating defaults. LGTM

File: src/main/Config.cpp [hunk 6]
```diff
@@ -1514,50 +1517,6 @@ Config::processConfig(std::shared_ptr<cpptoml::table> t)
             LIMIT_TX_QUEUE_SOURCE_ACCOUNT = true;
         }
 #endif
-        // PEER_FLOOD_READING_CAPACITY_BYTES (C): This is the initial credit
```
Comment (mbsdf) [inline src/main/Config.cpp:1517]: Can we update these comments to reflect new changes if any? I think these comments were useful.

File: src/main/Config.cpp [hunk 7]
```diff
@@ -1514,50 +1517,6 @@ Config::processConfig(std::shared_ptr<cpptoml::table> t)
...
-
-        // Receiver processes received bytes and once
-        // FLOW_CONTROL_SEND_MORE_BATCH_SIZE_BYTES (A) or more is processed, an
-        // acknowledgement is sent, which replenishes the sender's credit
-        // Sender (C-M1+A) <-- A bytes ---------- Receiver
-        //             \--- (C-M1+A)-M2 --->/
-
-        // Note:  M1, M2... are message sizes such that M <=
-        // MAX_CLASSIC_TX_SIZE_BYTES
-        if (!(PEER_FLOOD_READING_CAPACITY_BYTES -
```
Comment (mbsdf) [inline src/main/Config.cpp:1550]: this condition does not apply anymore, correct?

File: src/main/Config.cpp (hunk 6 above)
Comment (mbsdf) [reply src/main/Config.cpp:1517]: nvm

**Review by mbsdf — COMMENTED** [review]
couple of comments and questions

File: src/main/Config.cpp (hunk 7 above)
Comment (marta-lokhova) [reply src/main/Config.cpp:1550]: It still applies, but this check got moved to `HerderImpl::start` because we now need to load Soroban config to verify this condition (and we can't load the config until we start the application and initialize the database)

File: src/herder/HerderImpl.cpp (hunk 4 above)
Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:2006]: If operator relies on default (automatically calculated) values, no action is needed on restart. If the operator has custom values configured, at upgrade time, they'll be automatically adjusted to keep core running. However, on restart core will crash due to invalid parameters, so the operator will need to manually update them.

File: src/herder/HerderImpl.cpp [hunk 8]
```diff
@@ -1934,9 +1950,69 @@ HerderImpl::restoreUpgrades()
...
+            diff = conf.txMaxSizeBytes() - mMaxTxSize;
+        }
+        mMaxTxSize =
+            std::max<uint64_t>(getMaxClassicTxSize(), conf.txMaxSizeBytes());
+    }
+
+    // Maybe update capacity to reflect the upgrade
+    for (auto& peer : mApp.getOverlayManager().getAuthenticatedPeers())
+    {
+        peer.second->handleMaxTxSizeIncrease(diff);
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:1975]: yes, the node sends an additional SEND_MORE message to indicate an increase in capacity.

File: src/herder/HerderImpl.cpp (hunk 1 above)
Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:1968]: Are you referring to this? https://github.com/stellar/stellar-core/pull/3806/files#diff-d37d5bbb43097772a9cbca1b1215a7599e1fdd7be78ffb575fdc309dc9cf93d6R420 This PR already updates all places where getMaxTxSize should be used.

File: src/herder/HerderImpl.h (hunk 3 above)
Comment (marta-lokhova) [reply src/herder/HerderImpl.h:141]: is there an action item?

File: src/overlay/FlowControlCapacity.cpp [hunk 9]
```diff
@@ -97,6 +100,15 @@ FlowControlByteCapacity::canRead() const
     return true;
 }
 
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+void
+FlowControlByteCapacity::handleTxSizeIncrease(uint64_t increase)
+{
+    mCapacity.mFloodCapacity += increase;
```
Comment (SirTyson) [inline src/overlay/FlowControlCapacity.cpp:107]: Do we need to check for overflow here since this increase value is coming from another node and could potentially be malicious?

File: src/overlay/OverlayManagerImpl.h [hunk 10]
```diff
@@ -227,5 +228,6 @@ class OverlayManagerImpl : public OverlayManager
     void extractPeersFromMap(std::map<NodeID, Peer::pointer> const& peerMap,
                              std::vector<Peer::pointer>& result);
     void shufflePeerList(std::vector<Peer::pointer>& peerList);
+    std::pair<uint32_t, uint32_t> getFlowControlBytesConfig() const override;
```
Comment (SirTyson) [inline src/overlay/OverlayManagerImpl.h:231]: Nit: This std::pair interface is a bit hard to follow. Maybe break this into a struct or typedef? Or at least add a comment on the return type.

File: src/herder/HerderImpl.cpp [hunk 11]
```diff
@@ -1934,9 +1950,69 @@ HerderImpl::restoreUpgrades()
...
+    auto sz = mApp.getHerder().getMaxClassicTxSize();
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+    {
+        LedgerTxn ltx(mApp.getLedgerTxnRoot(),
+                      /* shouldUpdateLastModified */ true,
+                      TransactionMode::READ_ONLY_WITHOUT_SQL_TXN);
+        auto conf = mApp.getLedgerManager().getSorobanNetworkConfig(ltx);
+        mMaxTxSize = std::max<uint64_t>(sz, conf.txMaxSizeBytes());
+    }
+#else
```
Comment (SirTyson) [inline src/herder/HerderImpl.cpp:1992]: Nit: IDK if this is valid or not but I'm not a huge fan of the `#ifdef #else` value assignment. If you change 1983 to `mMaxTxSize = mApp.getHerder().getMaxClassicTxSize();` you can get rid of the `#else`.

**Review by SirTyson — COMMENTED** [review]
Overall looks good, just a few minor cleanups and one potential security issue.

File: src/overlay/FlowControlCapacity.cpp (hunk 9 above)
Comment (marta-lokhova) [reply src/overlay/FlowControlCapacity.cpp:107]: so here the increase is coming from the local node's upgrade logic, which I assume won't overflow (if the network upgrades to those types of values, then probably something is really wrong anyway). We do send another SEND_MORE message with `increase` in it, but we already handle potential overflow here (the mitigation is that the problematic peer is dropped) https://github.com/stellar/stellar-core/blob/0eab126b4343cf7d93af2cc9ca97c5bf7728ee4a/src/overlay/FlowControl.cpp#L340

File: src/herder/HerderImpl.cpp [hunk 12]
```diff
@@ -1934,9 +1950,69 @@ HerderImpl::restoreUpgrades()
     }
 }
 
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+void
+HerderImpl::maybeHandleUpgrade()
+{
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:1956]: mmm, `getSorobanNetworkConfig` doesn't fail if protocolVersion < 20, but instead creates an empty config, right? 
https://github.com/stellar/stellar-core/blob/0eab126b4343cf7d93af2cc9ca97c5bf7728ee4a/src/ledger/LedgerManagerImpl.cpp#L1148

So I believe this function should be a no-op if we're running protocol 19 (as confirmed by "upgrade to generalized tx set in network" test which runs the network at proto 19 for some time, then upgrades to 20). That being said, it does seem error-prone (e.g. what if this logic handles default values incorrectly), so I agree we can add a protocol check just in case.

File: src/overlay/FlowControlCapacity.cpp [hunk 13]
```diff
@@ -97,6 +100,15 @@ FlowControlByteCapacity::canRead() const
...
+    mCapacityLimits.mFloodCapacity += increase;
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/overlay/FlowControlCapacity.cpp:108]: discussed offline, but re-iterating here for posterity: one is node's current capacity, and the other is node's capacity limit.

File: src/overlay/FlowControl.cpp [hunk 14]
```diff
@@ -94,7 +94,7 @@ FlowControl::start(std::weak_ptr<Peer> peer,
         mFlowControlBytesCapacity =
             std::make_shared<FlowControlByteCapacity>(mApp, mNodeID);
         sendSendMore(mApp.getConfig().PEER_FLOOD_READING_CAPACITY,
-                     mApp.getConfig().PEER_FLOOD_READING_CAPACITY_BYTES,
+                     mApp.getOverlayManager().getFlowControlBytesConfig().first,
```
[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/overlay/FlowControl.cpp:97]: yeah, exactly

File: src/overlay/OverlayManagerImpl.h (hunk 10 above)
Comment (marta-lokhova) [reply src/overlay/OverlayManagerImpl.h:231]: good idea, using pair of the same type is also kind of error-prone. switched to a struct

File: src/overlay/FlowControlCapacity.cpp (hunk 9 above)
Comment (SirTyson) [reply src/overlay/FlowControlCapacity.cpp:107]: Looks good. Sorry, didn't realize this overflow check was already there!

File: src/herder/HerderImpl.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/herder/HerderImpl.cpp:1956]: Yeah it's probably fine as is, I just get a little spooked calling anything with Soroban in the name pre protocol 20.

[the remark this answers is not in the document]
Comment (marta-lokhova) [reply src/herder/HerderImpl.cpp:1956]: yeah, I also opened https://github.com/stellar/stellar-core/issues/3828 to prohibit retrieving soroban config prior to protocol 20

**Comment by marta-lokhova** [comment]
@SirTyson @mbsdf I believe the comments should be addressed now. Let me know if the change looks good.

**Review by mbsdf — APPROVED** [review]
LGTM

File: src/herder/HerderImpl.cpp [hunk 15]
```diff
@@ -1959,12 +1959,20 @@ HerderImpl::maybeHandleUpgrade()
         LedgerTxn ltx(mApp.getLedgerTxnRoot(),
                       /* shouldUpdateLastModified */ true,
                       TransactionMode::READ_ONLY_WITHOUT_SQL_TXN);
-        auto conf = mApp.getLedgerManager().getSorobanNetworkConfig(ltx);
+        if (ltx.loadHeader().current().ledgerVersion <
```
Comment (SirTyson) [inline src/herder/HerderImpl.cpp:1962]: Nit: Why not `protocolVersionIsBefore`?

**Review by SirTyson — APPROVED** [review]
LGTM

**Comment by MonsieurNicolas** [comment]
r+ 09c4774016a81a79b6d67022c7ae01709d21d29d
